src/de.rs

```diff
@@ -117,965 +70,516 @@ where
 /// assert_eq!(value, "foobar");
 /// ```
 #[cfg(feature = "std")]
-pub fn from_reader<T, R>(reader: R) -> Result<T>
+pub fn from_reader<T, R>(reader: R) -> Result<T, DecodeError<std::io::Error>>
 where
     T: de::DeserializeOwned,
-    R: io::Read,
+    R: std::io::BufRead,
 {
-    let mut deserializer = Deserializer::from_reader(reader);
-    let value = de::Deserialize::deserialize(&mut deserializer)?;
+    let reader = IoReader::new(reader);
+    let mut deserializer = Deserializer::new(reader);
+    let value = serde::Deserialize::deserialize(&mut deserializer)?;
     deserializer.end()?;
     Ok(value)
 }
 
-/// A Serde `Deserialize`r of CBOR data.
-#[derive(Debug)]
-pub struct Deserializer<R> {
-    read: R,
-    remaining_depth: u8,
-    accept_named: bool,
-    accept_packed: bool,
-    accept_standard_enums: bool,
-    accept_legacy_enums: bool,
+/// A Serde `Deserialize`r of DAG-CBOR data.
+struct Deserializer<R> {
+    reader: R,
 }
 
-#[cfg(feature = "std")]
-impl<R> Deserializer<IoRead<R>>
-where
-    R: io::Read,
-{
+impl<R> Deserializer<R> {
     /// Constructs a `Deserializer` which reads from a `Read`er.
-    pub fn from_reader(reader: R) -> Deserializer<IoRead<R>> {
-        Deserializer::new(IoRead::new(reader))
-    }
-}
-
-#[cfg(any(feature = "std", feature = "alloc"))]
-impl<'a> Deserializer<SliceRead<'a>> {
-    /// Constructs a `Deserializer` which reads from a slice.
-    ///
-    /// Borrowed strings and byte slices will be provided when possible.
-    pub fn from_slice(bytes: &'a [u8]) -> Deserializer<SliceRead<'a>> {
-        Deserializer::new(SliceRead::new(bytes))
-    }
-}
-
-impl<'a> Deserializer<MutSliceRead<'a>> {
-    /// Constructs a `Deserializer` which reads from a mutable slice that doubles as its own
-    /// scratch buffer.
-    ///
-    /// Borrowed strings and byte slices will be provided even for indefinite strings.
-    pub fn from_mut_slice(bytes: &'a mut [u8]) -> Deserializer<MutSliceRead<'a>> {
-        Deserializer::new(MutSliceRead::new(bytes))
-    }
-}
-
-impl<'a, 'b> Deserializer<SliceReadFixed<'a, 'b>> {
-    #[doc(hidden)]
-    pub fn from_slice_with_scratch(
-        bytes: &'a [u8],
-        scratch: &'b mut [u8],
-    ) -> Deserializer<SliceReadFixed<'a, 'b>> {
-        Deserializer::new(SliceReadFixed::new(bytes, scratch))
+    pub fn new(reader: R) -> Deserializer<R> {
```

Comment:
why not keep this as `from_reader`?

Reply:
Symmetry to the serializer. But as this is a public API, there's no point of breaking it.